enable calling SaltSSH minions from the salt master

gtmanfred opened the pull request:
### What does this PR do?
This adds the ability to let the salt-master spin off a new process which starts an SSHClient, and then returns the data back for the `salt` command to print it out.

### What issues does this PR fix or reference?
Closes #40997

### Tests written?

No

### Commits signed with GPG?

Yes

gtmanfred committed:
allow calling ssh minions from salt master process

gtmanfred committed:
add enable_ssh documentation

gtmanfred committed:
only fire events for stuff that is dictionaries

gtmanfred committed:
fix option name

thatch45 commented on salt/master.py:
```diff
@@ -1909,6 +1910,14 @@ def publish(self, clear_load):
+            ssh_minions = ssh._prep_ssh(**clear_load).targets.keys()
+            if ssh_minions:
+                minions.extend(ssh_minions)
+                multiprocessing.Process(target=ssh.cmd, kwargs=clear_load).start()
```
@s0undt3ch This process will go zombie if I am not mistaken. Can the process manager be modified to auto clean up a process when it dies instead of restarting it?

gtmanfred replied:
Just checked, and it does become a zombie.

gtmanfred committed:
dont pass syndic job lookups to ssh

gtmanfred commented:
Once kitchen multinode is a thing, this would be another great use case for tests.

It looks like this is going to require a lot of mocks added to get it working with the test suite.

gtmanfred committed:
this makes this all work with syndics

Any help would be great here, for some reason the jid gets updated and changed
somewhere in here, but only when using a syndic.

gtmanfred commented:
Any help on the last commit would be great, I can't figure out what is updating the jid in the `data` object, but if I just save it, and then reapply it, this makes the whole thing work with salt syndics as well.

gtmanfred committed:
remove old event that never got ran

gtmanfred committed:
make the jid in the ssh minion payload the same as the one for the tag

gtmanfred committed:
make sure all args are strings

when calling from `enable_ssh_minions` they could be other things.

gtmanfred committed:
Daemonize salt-ssh process

gtmanfred commented on salt/master.py:
```diff
@@ -1909,6 +1910,18 @@ def publish(self, clear_load):
+                def wrap_ssh(**kwargs):
+                    salt.utils.process.daemonize(False)
+                    ssh.cmd(**kwargs)
+                salt.utils.process.SignalHandlingMultiprocessingProcess(target=wrap_ssh, kwargs=clear_load).start()
```
This process is still becoming a zombie, even if it is daemonized, whatever runs the ssh connection is causing it to be zombied, possibly because of the generator that is created in `salt.client.ssh`, but if this is switched to threading the whole problem goes away.

@cachedout can you take a look.  Tom and I looked for a couple hours on Friday but couldn't figure out where to fix this.  We have the suspicion that the same thing also happens when running `salt-ssh` from the commandline, just that the zombie is cleaned up in a matter of milliseconds by the shell.

Thanks,
Daniel

gtmanfred commented:
Go Go Jenkins!

gtmanfred committed:
fix pylint

gtmanfred committed:
switch to threading

rallytime commented:
@gtmanfred Lint, please: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/16941/violations/file/salt/master.py/

gtmanfred commented:
bah, fixed

On Tue, Dec 5, 2017 at 1:41 PM, Nicole Thomas <notifications@github.com>
wrote:

> @gtmanfred <https://github.com/gtmanfred> Lint, please:
> https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/
> 16941/violations/file/salt/master.py/
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/44578#issuecomment-349434953>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAssoT28wE9TN3kbalozIqO-eIFrA3c-ks5s9aqFgaJpZM4QhLfD>
> .
>

rallytime commented:
@gtmanfred Unfortunately, there's a merge conflict now. Can you rebase?

gtmanfred commented:
done

On Thu, Dec 7, 2017 at 2:27 PM, Nicole Thomas <notifications@github.com>
wrote:

> @gtmanfred <https://github.com/gtmanfred> Unfortunately, there's a merge
> conflict now. Can you rebase?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/saltstack/salt/pull/44578#issuecomment-350099429>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAssoWvZh93ClAojNDzQI3Jdf8LJ5I3-ks5s-FhCgaJpZM4QhLfD>
> .
>

rallytime commented:
Ah, sorry @gtmanfred - this will need another rebase.

cachedout commented on salt/client/ssh/__init__.py:
```diff
@@ -663,11 +663,34 @@ def run_iter(self, mine=False, jid=None):
+                _, data = next(six.iteritems(ret))
+                data['jid'] = jid  # make the jid in the payload the same as the jid in the tag
                 self.event.fire_event(
+<<<<<<< HEAD
```
You have unresolved merge issues in here.

gtmanfred commented:
Yup, missed that one, just finished rebasing.

gtmanfred commented:
Go Go Jenkins!

rallytime commented:
Oh, unfortunately this has another merge conflict. Can you rebase one more time @gtmanfred?

gtmanfred commented:
Done

rallytime commented:
@gtmanfred There's some lint here: https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/17375/violations/file/salt/config/__init__.py/

gtmanfred commented:
Should be good now.

cachedout commented on doc/ref/configuration/master.rst:
```diff
@@ -96,6 +96,23 @@ The user to run the Salt processes
+
+Default: ``False``
+
+Tell the master to also use SaltSSH when running commands against minions.
```
This has always been `salt-ssh` and not `SaltSSH`. (Though I do like that. If we change it we should change it everywhere, however.)

cachedout commented on doc/ref/configuration/master.rst:
```diff
@@ -96,6 +96,23 @@ The user to run the Salt processes
+
+.. note::
+
+    Cross minion type communication is still not possible.  The SaltMine and
```
`Cross-minion`. Probably remove the word `type` here.

cachedout replied:
`Salt mine`. Not `SaltMine`.

cachedout commented on salt/client/ssh/__init__.py:
```diff
@@ -215,7 +215,7 @@ class SSH(object):
         self.__parsed_rosters = {SSH.ROSTER_UPDATE_FLAG: True}
         pull_sock = os.path.join(opts['sock_dir'], 'master_event_pull.ipc')
-        if os.path.isfile(pull_sock) and HAS_ZMQ:
+        if os.path.exists(pull_sock) and HAS_ZMQ:
```
Hmm, that's interesting. Why was this change necessary?

gtmanfred replied:
So, this never actually worked, os.path.isfile does not return true on sockets, so we were never firing these events on the event bus in the past.

That is why I just threw away the old event that would be thrown and changed it to be the same format as our regular minion events.

cachedout commented on salt/master.py:
```diff
@@ -1957,6 +1959,15 @@ def publish(self, clear_load):
         payload = self._prep_pub(minions, jid, clear_load, extra, missing)
 
         # Send it!
+        if self.opts[u'enable_ssh_minions'] is True and isinstance(clear_load[u'tgt'], six.string_types):
```
This method (`publish`) has so much bloat, I think it would be best to split this out into a private method.

gtmanfred committed:
split out publishing to ssh minions

and clean up docs

rallytime commented:
@gtmanfred Lint :) https://jenkins.saltstack.com/job/PR/job/salt-pr-lint-n/17474/violations/file/salt/master.py/

gtmanfred committed:
fix pylint

rallytime commented:
@gtmanfred One more rebase here and then we should be able to get this in. :)

gtmanfred commented:
:+1: done
